src/buffer_manager.ts

```diff
@@ -410,6 +412,74 @@ export class BufferManager implements Disposable {
 
     private handleWindowChangedDebounced = debounce(this.handleWindowChanged, 100, { leading: false, trailing: true });
 
+    private handleSaveBuf(bufID: number, filename: string) {
+        const document = this.getTextDocumentForBufferId(bufID);
+        if (document == null) {
+            throw new Error("Received an invalid buffer id from Neovim, cannot save");
+        }
+
+        this.saveDocument(bufID, document, filename).catch((err) => {
+            window.showErrorMessage(`Failed to save document to ${filename}: ${err}`);
+        });
+    }
+
+    private async saveDocument(bufID: number, document: TextDocument, filename: string): Promise<void> {
+        const needSaveAs = await this.needSaveAs(document, filename);
+        if (needSaveAs) {
+            const newURI = await workspace.saveAs(document.uri);
+            if (newURI === undefined) {
+                throw new Error("Unknown document save failure");
+            }
+
+            const bufname = await this.bufnameForURI(newURI);
+            await this.client.lua(
+                `
+                local bufId, docUri, docUriJson, bufname = ...
+
+                vim.api.nvim_buf_set_var(bufId, "vscode_uri", docUri)
+                vim.api.nvim_buf_set_var(bufId, "vscode_uri_data", docUriJson)
+                vim.api.nvim_buf_set_name(bufId, bufname)
+                vim.bo[bufId].modified = false
+            `,
+                [bufID, document.uri.toString(), document.uri.toJSON(), bufname],
+            );
```

Comment:
```suggestion
            await workspace.saveAs(document.uri);
```

We just need to call the method, the document management is handled by `syncEditorLayout`.